File: plugins/examples/async_demo.lua
-- Async Process Demo Plugin
-- Demonstrates spawning external processes asynchronously

editor.register_command({
    name = "Async Demo: Git Status",
    description = "Run git status and show output",
    action = "async_git_status",
    contexts = {"normal"},
    callback = function()
        editor.set_status("Running git status...")

        editor.spawn("git", {"status", "--short"}, function(stdout, stderr, exit_code)
            if exit_code == 0 then
                if stdout == "" or stdout == "\n" then
                    editor.set_status("Git: Working tree clean")
                else
                    -- Count lines
                    local count = 0
                    for _ in stdout:gmatch("\n") do count = count + 1 end
                    editor.set_status(string.format("Git: %d files changed", count))
                end
            else
                editor.set_status("Git status failed: " .. stderr)
            end
        end)
    end
})

editor.register_command({
    name = "Async Demo: Current Directory",
    description = "Show current directory using pwd",
    action = "async_pwd",
    contexts = {"normal"},
    callback = function()
        editor.spawn("pwd", {}, function(stdout, stderr, exit_code)
            if exit_code == 0 then
                local dir = stdout:gsub("\n", "")
                editor.set_status("Current directory: " .. dir)
            else
                editor.set_status("pwd failed")
            end
        end)
    end
})

editor.register_command({
    name = "Async Demo: List Files",
    description = "List files in current directory",
    action = "async_ls",
    contexts = {"normal"},
    callback = function()
        editor.set_status("Listing files...")

        editor.spawn("ls", {"-1"}, function(stdout, stderr, exit_code)
            if exit_code == 0 then
                local count = 0
                for _ in stdout:gmatch("\n") do count = count + 1 end
                editor.set_status(string.format("Found %d files/directories", count))
            else
                editor.set_status("ls failed")
            end
        end)
    end
})

editor.register_command({
    name = "Async Demo: Git Branch",
    description = "Show current git branch",
    action = "async_git_branch",
    contexts = {"normal"},
    callback = function()
        editor.spawn("git", {"branch", "--show-current"}, function(stdout, stderr, exit_code)
            if exit_code == 0 then
                local branch = stdout:gsub("\n", "")
                if branch ~= "" then
                    editor.set_status("Git branch: " .. branch)
                else
                    editor.set_status("Not on any branch (detached HEAD)")
                end
            else
                editor.set_status("Not a git repository")
            end
        end)
    end
})

editor.register_command({
    name = "Async Demo: Echo Test",
    description = "Test with simple echo command",
    action = "async_echo",
    contexts = {"normal"},
    callback = function()
        editor.spawn("echo", {"Hello from async process!"}, function(stdout, stderr, exit_code)
            editor.set_status("Echo output: " .. stdout:gsub("\n", ""))
        end)
    end
})

editor.register_command({
    name = "Async Demo: With Working Dir",
    description = "Run command in /tmp directory",
    action = "async_with_cwd",
    contexts = {"normal"},
    callback = function()
        editor.spawn("pwd", {}, {cwd = "/tmp"}, function(stdout, stderr, exit_code)
            local dir = stdout:gsub("\n", "")
            editor.set_status("Working dir was: " .. dir)
        end)
    end
})

editor.register_command({
    name = "Async Demo: Error Handling",
    description = "Demonstrate error handling with non-existent command",
    action = "async_error",
    contexts = {"normal"},
    callback = function()
        editor.spawn("this_command_does_not_exist", {}, function(stdout, stderr, exit_code)
            if exit_code ~= 0 then
                editor.set_status("Command failed (as expected): " .. stderr)
            else
                editor.set_status("Unexpected success")
            end
        end)
    end
})

editor.set_status("Async Demo plugin loaded! Try the 'Async Demo' commands.")

File: plugins/examples/async_demo.ts
/// <reference path="../../types/fresh.d.ts" />

/**
 * Async Process Demo Plugin
 * Demonstrates spawning external processes asynchronously with async/await
 */

// Git status
globalThis.async_git_status = async function(): Promise<void> {
  editor.setStatus("Running git status...");

  try {
    const result = await editor.spawnProcess("git", ["status", "--short"]);
    if (result.exit_code === 0) {
      if (result.stdout === "" || result.stdout === "\n") {
        editor.setStatus("Git: Working tree clean");
      } else {
        const count = result.stdout.split("\n").filter(line => line.trim()).length;
        editor.setStatus(`Git: ${count} files changed`);
      }
    } else {
      editor.setStatus(`Git status failed: ${result.stderr}`);
    }
  } catch (e) {
    editor.setStatus(`Git status error: ${e}`);
  }
};

editor.registerCommand(
  "Async Demo: Git Status",
  "Run git status and show output",
  "async_git_status",
  "normal"
);

// Current directory
globalThis.async_pwd = async function(): Promise<void> {
  try {
    const result = await editor.spawnProcess("pwd");
    if (result.exit_code === 0) {
      const dir = result.stdout.trim();
      editor.setStatus(`Current directory: ${dir}`);
    } else {
      editor.setStatus("pwd failed");
    }
  } catch (e) {
    editor.setStatus(`pwd error: ${e}`);
  }
};

editor.registerCommand(
  "Async Demo: Current Directory",
  "Show current directory using pwd",
  "async_pwd",
  "normal"
);

// List files
globalThis.async_ls = async function(): Promise<void> {
  editor.setStatus("Listing files...");

  try {
    const result = await editor.spawnProcess("ls", ["-1"]);
    if (result.exit_code === 0) {
      const count = result.stdout.split("\n").filter(line => line.trim()).length;
      editor.setStatus(`Found ${count} files/directories`);
    } else {
      editor.setStatus("ls failed");
    }
  } catch (e) {
    editor.setStatus(`ls error: ${e}`);
  }
};

editor.registerCommand(
  "Async Demo: List Files",
  "List files in current directory",
  "async_ls",
  "normal"
);

// Git branch
globalThis.async_git_branch = async function(): Promise<void> {
  try {
    const result = await editor.spawnProcess("git", ["branch", "--show-current"]);
    if (result.exit_code === 0) {
      const branch = result.stdout.trim();
      if (branch !== "") {
        editor.setStatus(`Git branch: ${branch}`);
      } else {
        editor.setStatus("Not on any branch (detached HEAD)");
      }
    } else {
      editor.setStatus("Not a git repository");
    }
  } catch (e) {
    editor.setStatus(`Git branch error: ${e}`);
  }
};

editor.registerCommand(
  "Async Demo: Git Branch",
  "Show current git branch",
  "async_git_branch",
  "normal"
);

// Echo test
globalThis.async_echo = async function(): Promise<void> {
  try {
    const result = await editor.spawnProcess("echo", ["Hello from async process!"]);
    editor.setStatus(`Echo output: ${result.stdout.trim()}`);
  } catch (e) {
    editor.setStatus(`Echo error: ${e}`);
  }
};

editor.registerCommand(
  "Async Demo: Echo Test",
  "Test with simple echo command",
  "async_echo",
  "normal"
);

// With working directory
globalThis.async_with_cwd = async function(): Promise<void> {
  try {
    const result = await editor.spawnProcess("pwd", [], "/tmp");
    const dir = result.stdout.trim();
    editor.setStatus(`Working dir was: ${dir}`);
  } catch (e) {
    editor.setStatus(`pwd error: ${e}`);
  }
};

editor.registerCommand(
  "Async Demo: With Working Dir",
  "Run command in /tmp directory",
  "async_with_cwd",
  "normal"
);

// Error handling
globalThis.async_error = async function(): Promise<void> {
  try {
    const result = await editor.spawnProcess("this_command_does_not_exist");
    if (result.exit_code !== 0) {
      editor.setStatus(`Command failed (as expected): exit ${result.exit_code}`);
    } else {
      editor.setStatus("Unexpected success");
    }
  } catch (e) {
    editor.setStatus(`Command failed with error: ${e}`);
  }
};

editor.registerCommand(
  "Async Demo: Error Handling",
  "Demonstrate error handling with non-existent command",
  "async_error",
  "normal"
);

editor.setStatus("Async Demo plugin loaded! Try the 'Async Demo' commands.");
editor.debug("Async Demo plugin initialized with native async/await support");
